File: DMD/DMDPDF/DMDPDF.samples/samples/cli/AttachmentExtractionSample.vb
Imports DMD.org.dmdpdf.bytes
Imports DMD.org.dmdpdf.documents
Imports DMD.org.dmdpdf.documents.files
Imports DMD.org.dmdpdf.documents.interaction.annotations
Imports DMD.org.dmdpdf.files
Imports DMD.org.dmdpdf.objects

Imports System
Imports System.Collections.Generic
Imports System.IO

Namespace org.dmdpdf.samples.cli

    '/**
    '  <summary>This sample demonstrates how to extract attachments from a PDF document.</summary>
    '*/
    Public Class AttachmentExtractionSample
        Inherits Sample

        Public Overrides Sub Run()
            ' 1. Opening the PDF file...
            Dim filePath As String = PromptFileChoice("Please select a PDF file")
            Using file As New DMD.org.dmdpdf.files.File(filePath)
                Dim document As Document = file.Document

                ' 2. Extracting attachments...
                ' 2.1. Embedded files (document level).
                For Each entry As KeyValuePair(Of PdfString, FileSpecification) In document.Names.EmbeddedFiles
                    EvaluateDataFile(entry.Value)
                Next

                ' 2.2. File attachments (page level).
                For Each page As Page In document.Pages
                    For Each annotation As Annotation In page.Annotations
                        If (TypeOf (annotation) Is FileAttachment) Then
                            EvaluateDataFile(CType(annotation, FileAttachment).DataFile)
                        End If
                    Next
                Next
            End Using
        End Sub

        Private Sub EvaluateDataFile(ByVal dataFile As FileSpecification)
            If (TypeOf (dataFile) Is FullFileSpecification) Then
                Dim embeddedFile As EmbeddedFile = CType(dataFile, FullFileSpecification).EmbeddedFile
                If (embeddedFile IsNot Nothing) Then
                    ExportAttachment(embeddedFile.Data, dataFile.Path)
                End If
            End If
        End Sub

        Private Sub ExportAttachment(ByVal data As IBuffer, ByVal filename As String)
            Dim outputPath As String = GetOutputPath(filename)
            Dim outputStream As FileStream
#If Not DEBUG Then
            Try
#End If
            outputStream = New FileStream(outputPath, FileMode.CreateNew)
#If Not DEBUG Then
            Catch e As Exception
                Throw New Exception(outputPath & " file couldn't be created.", e)
            End Try
#End If

#If Not DEBUG Then
            Try
#End If
            Dim writer As BinaryWriter = New BinaryWriter(outputStream)
            writer.Write(data.ToByteArray())
            writer.Close()
            outputStream.Close()
#If Not DEBUG Then
            Catch e As Exception
                Throw New Exception(outputPath & " file writing has failed.", e)
            End Try
#End If
            Me.OutputLine("Output: " & outputPath)
        End Sub

    End Class
End Namespace

File: DMD/DMDPDF/DMDPDF.samples/samples/cli/AttachmentExtractionSample.cs
using org.dmdpdf.bytes;
using org.dmdpdf.documents;
using org.dmdpdf.documents.files;
using org.dmdpdf.documents.interaction.annotations;
using files = org.dmdpdf.files;
using org.dmdpdf.objects;

using System;
using System.Collections.Generic;
using System.IO;

namespace org.dmdpdf.samples.cli
{
  /**
    <summary>This sample demonstrates how to extract attachments from a PDF document.</summary>
  */
  public class AttachmentExtractionSample
    : Sample
  {
    public override void Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptFileChoice("Please select a PDF file");
      using(files::File file = new files::File(filePath))
      {
        Document document = file.Document;

        // 2. Extracting attachments...
        // 2.1. Embedded files (document level).
        foreach(KeyValuePair<PdfString,FileSpecification> entry in document.Names.EmbeddedFiles)
        {EvaluateDataFile(entry.Value);}

        // 2.2. File attachments (page level).
        foreach(Page page in document.Pages)
        {
          foreach(Annotation annotation in page.Annotations)
          {
            if(annotation is FileAttachment)
            {EvaluateDataFile(((FileAttachment)annotation).DataFile);}
          }
        }
      }
    }

    private void EvaluateDataFile(
      FileSpecification dataFile
      )
    {
      if(dataFile is FullFileSpecification)
      {
        EmbeddedFile embeddedFile = ((FullFileSpecification)dataFile).EmbeddedFile;
        if(embeddedFile != null)
        {ExportAttachment(embeddedFile.Data, dataFile.Path);}
      }
    }

    private void ExportAttachment(
      IBuffer data,
      string filename
      )
    {
      string outputPath = GetOutputPath(filename);
      FileStream outputStream;
      try
      {outputStream = new FileStream(outputPath, FileMode.CreateNew);}
      catch(Exception e)
      {throw new Exception(outputPath + " file couldn't be created.",e);}

      try
      {
        BinaryWriter writer = new BinaryWriter(outputStream);
        writer.Write(data.ToByteArray());
        writer.Close();
        outputStream.Close();
      }
      catch(Exception e)
      {throw new Exception(outputPath + " file writing has failed.",e);}

      Console.WriteLine("Output: " + outputPath);
    }
  }
}